src/message-stream.ts

```diff
@@ -0,0 +1,366 @@
+    super(status.details);
+    this.code = status.code;
+    this.metadata = status.metadata;
+  }
+}
+
+/**
+ * @typedef {object} MessageStreamOptions
+ * @property {number} [highWaterMark=0] Configures the Buffer level for all
+ *     underlying streams. See {@link
+ * https://nodejs.org/en/docs/guides/backpressuring-in-streams/} for more
+ * details.
+ * @property {number} [maxStreams=5] Number of streaming connections to make.
+ * @property {number} [timeout=300000] Timeout for establishing a connection.
+ */
+export interface MessageStreamOptions {
+  highWaterMark: number;
+  maxStreams?: number;
+  timeout?: number;
+}
+
+/**
+ * Streaming class used to manage multiple StreamingPull requests.
+ *
+ * @private
+ * @class
+ *
+ * @param {Subscriber} sub The parent subscriber.
+ * @param {MessageStreamOptions} [options] The message stream options.
+ */
+export class MessageStream extends PassThrough {
+  destroyed: boolean;
+  private _filling: boolean;
+  private _keepAliveHandle: NodeJS.Timer;
+  private _options!: MessageStreamOptions;
+  private _streams: Set<GaxDuplex>;
+  private _subscriber: Subscriber;
+  constructor(sub: Subscriber, options = {} as MessageStreamOptions) {
+    const {highWaterMark = 0} = options;
+    super({objectMode: true, highWaterMark});
+
+    this.destroyed = false;
+    this._filling = false;
+    this._streams = new Set();
+    this._subscriber = sub;
+    this._keepAliveHandle =
+        setInterval(() => this._keepAlive(), KEEP_ALIVE_INTERVAL);
+
+    this._keepAliveHandle.unref();
+    this.setOptions(options);
+  }
+  /**
+   * @private
+   * @type {boolean}
+   */
+  private get _needsFill(): boolean {
+    return this._streams.size < this._options.maxStreams!;
+  }
+  /**
+   * Adds a StreamingPull stream to the combined stream.
+   *
+   * @param {stream} stream The StreamingPull stream.
+   */
+  add(stream: GaxDuplex): void {
+    stream.receivedStatus = false;
+
+    this._setHighWaterMark(stream);
+    this._streams.add(stream);
+
+    stream.on('error', err => this._onerror(stream, err))
+        .once('status', status => this._onstatus(stream, status))
+        .once('readable', () => this._setHighWaterMark(stream.stream))
+        .pipe(this, {end: false});
+  }
+  /**
+   * Destroys the stream and any underlying streams.
+   *
+   * @param {error?} err An error to emit, if any.
+   */
+  destroy(err?: Error): void {
+    if (this.destroyed) {
+      return;
+    }
+
+    this.destroyed = true;
+    clearInterval(this._keepAliveHandle);
+
+    this._streams.forEach(stream => {
+      this.remove(stream);
+      stream.cancel();
+    });
+
+    if (super.destroy) {
+      return super.destroy(err);
+    }
+
+    process.nextTick(() => {
+      if (err) {
+        this.emit('error', err);
+      }
+      this.emit('close');
+    });
+  }
+  /**
+   * Removes a stream from the combined stream.
+   *
+   * @param {stream} stream The stream to remove.
+   */
+  remove(stream: GaxDuplex): void {
```

Comment:
I think `removeSteam` could be more descriptive here. Otherwise, my instinct is that `MessageStream.remove()` would remove a "message".